astromon/db.py

```diff
@@ -408,13 +242,42 @@ def save(table_name, data, dbfile):
         File where tables are stored.
         The default is `$ASTROMON_FILE` or `$SKA/data/astromon/astromon.h5`
     """
-    with connect(dbfile) as con:
-        if type(con) is tables.file.File:
-            return _save_hdf5(con, table_name, data)
-        elif type(con) is sqlite3.Connection:
-            return _save_sql(con, table_name, data)
+    with connect(dbfile, mode='r+') as h5:
+        # sanity checks: assert that file is open for writing
+        assert h5.isopen, f'{h5.filename} is not open'
+        assert h5.mode in ['r+', 'w'], f'{h5.filename} is not open for writing'
+        assert isinstance(data, table.Table), 'input to _save_hdf5 must be a table'
+
+        if table_name in DTYPES:
+            dtype = DTYPES[table_name]
+            names = [n for n in dtype.names if n in data.dtype.names]
+            if len(names) == 0:
+                raise Exception('Input data has no columns in common with table in DB')
+            missing = [name for name in dtype.names if name not in data.dtype.names]
+            if missing:
+                raise Exception(
+                    f'Saving table {table_name} with missing columns: {", ".join(missing)}'
+                )
+            b = np.zeros(len(data), dtype=dtype)
+            b[names] = data[names].as_array().astype(dtype[names])
+            data = b
         else:
-            raise Exception(f'Unknown DB type: {dbfile}/{con}')
+            data = data.as_array()
+
+        if table_name in h5.root:
+            node = h5.get_node(f'/{table_name}')
+            if 'obsid' in data.dtype.names:
+                # remove rows for these obsids
+                obsids = np.unique(data['obsid'])
+                data_out = node[:].astype(dtype)
+                data_out = data_out[~np.in1d(data_out['obsid'], obsids)]
+                # append current data
+                data = np.concatenate((data_out, data))
+            h5.remove_node(node)
```

Comment:
I was curious if the strategy of removing nodes and re-creating could bloat the file, so I did an experiment of updating an HDF5 in this way three thousand times. Answer: it can make the file 10% larger, but not a factor of 10. That's good news and means this idiom can be re-used in general. Nice!